src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -0,0 +1,849 @@
+// user to convert them to actual qemu command line parameter strings.
+
+trait ToQemuParams {
+    // OsString could look as a better fit here, however since foreign strings
+    // come to this code from the outside as Strings already and this code adds
+    // nothing but UTF-8 (in fact probably just ASCII) switching to OsStrings
+    // now seems pointless.
+    fn qemu_params(&self) -> Result<Vec<String>>;
+}
+
+#[derive(Debug)]
+struct Kernel {
+    // PathBuf would seem more appropriae but since we get the kernel path
+    // from config as String already and we do no path operations on it,
+    // converting to PathBuf and then back to String seems futile
+    path: String,
+    initrd_path: String,
+    params: KernelParams,
+}
+
+impl Kernel {
+    fn new(config: &HypervisorConfig) -> Result<Kernel> {
+        // get kernel params
+        let mut kernel_params = KernelParams::new(config.debug_info.enable_debug);
+
+        if config.boot_info.initrd.is_empty() {
+            // QemuConfig::validate() has already made sure that if initrd is
+            // empty, image cannot be so we don't need to re-check that here
+
+            kernel_params.append(
+                &mut KernelParams::new_rootfs_kernel_params(
+                    &config.boot_info.vm_rootfs_driver,
+                    &config.boot_info.rootfs_type,
+                )
+                .context("adding rootfs params failed")?,
+            );
+        }
+
+        kernel_params.append(&mut KernelParams::from_string(
+            &config.boot_info.kernel_params,
+        ));
+
+        Ok(Kernel {
+            path: config.boot_info.kernel.clone(),
+            initrd_path: config.boot_info.initrd.clone(),
+            params: kernel_params,
+        })
+    }
+}
+
+impl ToQemuParams for Kernel {
+    fn qemu_params(&self) -> Result<Vec<String>> {
+        let mut result = Vec::new();
+
+        // QemuConfig::adjust_config() ensures that kernel path is never empty
+        result.push("-kernel".to_owned());
+        result.push(self.path.clone());
+
+        if !self.initrd_path.is_empty() {
+            result.push("-initrd".to_owned());
+            result.push(self.initrd_path.clone());
+        }
+
+        let kernel_params = self.params.to_string()?;
+        if !kernel_params.is_empty() {
+            result.push("-append".to_owned());
+            result.push(kernel_params);
+        }
+
+        Ok(result)
+    }
+}
+
+fn format_memory(mem_size: u64) -> String {
+    if mem_size % GI_B == 0 {
+        format!("{}G", mem_size / GI_B)
+    } else if mem_size % MI_B == 0 {
+        format!("{}M", mem_size / MI_B)
+    } else {
+        format!("{}", mem_size)
+    }
+}
+
+#[derive(Debug)]
+struct Memory {
+    // 'size' and 'max_size' are stored in bytes
+    size: u64,
+    num_slots: u32,
+    max_size: u64,
+    memory_backend_file: Option<MemoryBackendFile>,
+}
+
+impl Memory {
+    fn new(config: &HypervisorConfig) -> Memory {
```

Comment:
I believe the bulk of the behaviour of this function comes from the golang shim which IIRC doesn't treat the conditions as errors (I think it doesn't even log them, I believe I chose to log them) so the constructor cannot actually fail.

My uncertainty and a question about this is rather, does the checking/adjusting part of this function actually belong to `QemuConfig::adjust_config()`? I kind of believe it does, if the values are sanitised right after parsing the config file the runtime proper will see the actual values which might help and shouldn't hurt. I haven't seen this kind of more extensive non-trivial activity in other `adjust_config()` implementations though so I wasn't sure where to put it.  What do you think?

Reply:
@pmores - imho we shouldn't be duplicating the golang approach "exactly" since runtime-rs is very different internally to the golang runtime.

Due to some of the complex scenarios we need to deal with, the approach we've taken in the CH driver is to check everything immediately and aggressively, and fail fast. That has the advantage of giving us assurances but also means we can test unit test these scenarios.

So, yes, the CH `adjust_config()` is fairly minimal and basically just sets default values.

Reply:
@jodh-intel Let me check if I understand correctly. `adjust_config()` is run very early in the process of launching the shim, right after loading the configuration, so checking immediately would favour putting the memory adjustment code there, right?  Also, it would have the advantage that the shim proper that launches after loading and adjusting the config would see the actual values used by the hypervisor (should they need that which I don't know if they do).  Based on that, I'd be inclined to move the adjustment code from here to `adjust_config()` although I'd like to do so in a follow-up PR so as not to complicate merging this one.